### ISSUE TITLE
Urls in labels should be easy to find

### ISSUE THREAD

**Issue by freeseacher** [VictoriaMetrics/VictoriaMetrics#532]
**Is your feature request related to a problem? Please describe.**
Sometimes after code refactoring labels will contain not what was expected since start. Once i got an urls in label "calling_service" instead of service name. And i got long url with get params in it. 
That was unexpected from both developer and admin side. Of cause that leads to large cardinality

**Describe the solution you'd like**
The main idea is to get an list with blacklisted chars in label value. Than count if value matchs agains it. Probably that should not apply to each insert operation but can be applied to some sampling subset. 

**Describe alternatives you've considered**
that operation can be applied on metric relabling during scrape operation too. But i can't do anything with it excerpt drop label or even metric
Also that operation can be done on select, but i really have no idea how enumerate all labels and find chars in it

**Comment by valyala** [comment]
Side note: it is possible to remove unsupported chars from label values with relabeling in [vmagent](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/app/vmagent/README.md) and single-node VictoriaMetrics with `replace_all` action. See [these docs](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/app/vmagent/README.md#relabeling) for details.

For example, the following relabeling rule replaces all the chars outside `a-zA-Z0-9` with `_` in the `calling_service` label:

```yml
- action: replace_all
  regex: "[^a-zA-Z0-9]"
  replacement: "_"
  source_labels: [calling_service]
  target_label: calling_service

**Comment by freeseacher** [comment]
yep it is. but it will not fight agains carnality. Only replaces values. The idea is to find that we probably have misconfig and then find where and fix it. not hide.

**Comment by valyala** [comment]
VictoriaMetrics exposes labels with the maximum number of label values at `labelValueCountByLabelName` array at [/api/v1/status/tsdb](https://prometheus.io/docs/prometheus/latest/querying/api/#tsdb-stats) endpoint. Probably this endpoint could be used for quick detecting of high cardinality labels. VictoriaMetrics accepts optional `date=YYYY-MM-DD` and `topN=42` args on this page. By default `date` equals to the current date, while `topN` equals to 10.

**Comment by freeseacher** [comment]
and again that was not so helpfull, i have several kubernetes clusters and of cause labels like `id` and `pod` and `service` has much more cardinality. 
by the way topN looks great and probably will help diagnose problem

**Comment by valyala** [comment]
FYI, take a look at https://github.com/open-fresh/bomb-squad (thanks to Ren from [slack community chat](http://slack.victoriametrics.com/) for the link).

**Comment by zbindenren** [comment]
Unfortunately it seams that the bomb squad project is dead. Would be nice if something similar is included in vm.

**Comment by freeseacher** [comment]
And more over should this issue have to be generalized to "fight against high carnality" or has separate value i'm not sure. 
For me this is a red flag that we have some issues for debug not the debug it self

**Comment by makasim** [comment]
I don’t think the proposed pattern-matching approach is sufficient. Problematic labels can contain values like timestamps or IDd that are indistinguishable from legitimate ones, making reliable filtering in the general case impossible.

A more robust approach is to monitor cardinality across selected dimensions and enforce quotas when limits are exceeded. These dimensions could include job, job + metric name, tenant, region, etc. Instead of trying to guess “bad” labels, this focuses on controlling the actual source of cardinality growth.

I’ve outlined this approach in a proposal: https://github.com/VictoriaMetrics/VictoriaMetrics/issues/10848. There’s also a proof of concept available here: https://github.com/makasim/cestimator.

Take a look and share your feedback.

**Comment by makasim** [comment]
I’d like to share an alternative approach: what if we could limit cardinality growth before it reaches storage without writing validation rules by hand. It would work in general case. 

For this, we could leverage [vmestimator](https://github.com/VictoriaMetrics/vmestimator). It can already measure cardinality across arbitrary label dimensions in real time. This functionality could be extended to support cardinality limits for selected dimensions.

For example:

```yaml
streams:
  - group_by: ['job']
    interval: '15m'

# to be implemented part
limiters:
  # Exception for the bar job
  - filter: '{job="bar"}'
    threshold_up: 10000
    threshold_down: 8000

  # Catch-all limiter
  - threshold_up: 5000
    threshold_down: 4000
```

When a particular job exceeds the configured limit, vmestimator could generate a relabeling rule to drop the affected series:

```yaml
- action: drop
  source_labels: [job]
  regex: "bar"
```

vmagent could periodically load the generated configuration from vmestimator. 

```
/path/to/vmagent \
  -remoteWrite.url=http://victoria-metrics:8428/api/v1/write \
  -remoteWrite.urlRelabelConfig=http://vmestimator:8490/relabel_config \
  -remoteWrite.urlRelabelConfigRefreshInterval=5m  # to be implemented
```

Once the cardinality drops below `threshold_down`, vmestimator would remove the drop rule. This would provide a fast and automated mitigation mechanism for cardinality explosions.

@freeseacher Would you use the proposed approach?

**Comment by makasim** [comment]
Closing as not planned
